### PR TITLE
Add diagonal matrix compilation routine from tweedledum

### PR DESCRIPTION
**Pull request by notmgsk**

### CODE REVIEW

File: src/contrib/tweedledum/tweedledum.lisp [hunk 1]
```diff
@@ -69,6 +69,43 @@ GIVE-UP-COMPILATION if INSTR is not a permutation gate."
...
+  ;; instruction. Like with permutation gates, diagonality should be
+  ;; available as part of the structure definition. Maybe that already
+  ;; exists.
+  (let ((m (quil:gate-matrix instr)))
+    (unless (diagonal-p m)
+      (quil::give-up-compilation))
+    ;; This synthesis routine works on the unitary
+    ;;   U = diag(1, e^{-i t_1}, ..., e^{-i t_{2^n - 1}})
+    ;; and takes as input the angles t_i; hence (rest ...) below.
+    (let ((angles (rest (mapcar #'phase (magicl:matrix-diagonal m)))))
```
Comment (jmbr) [inline src/contrib/tweedledum/tweedledum.lisp]: Suggestion: use `angles (mapcar #'phase (rest (magicl:matrix-diagonal m)))` instead (no need to evaluate `phase` on the constant 1 and then drop it, we can just drop it from the get go).

**Comment by jmbr** [comment]
This PR gives me a wonderful opportunity to raise https://github.com/rigetti/quilc/pull/305 back from the dead and bring it to your (@notmgsk and possibly @stylewarning 's) attention.

**Comment by notmgsk** [comment]
> This PR gives me a wonderful opportunity to raise #305 back from the dead and bring it to your (@notmgsk and possibly @stylewarning 's) attention.

ok ok ok :)

File: src/contrib/tweedledum/tweedledum.c [hunk 2]
```diff
@@ -14,6 +15,20 @@
...
+  extern char* tweedledum_synthesis_diagonal(double* angles, uint32_t size) {
+    using namespace tweedledum;
+    std::vector<double> vangles(angles, angles + size);
+    std::cout << vangles.size() << std::endl;
+    auto network = diagonal_synth<gg_network<mcmt_gate>>(vangles);
+
+    std::stringstream ss;
+    write_quil(network, ss);
+    const std::string resultstr = ss.str();
+    char* str = (char*) malloc( (resultstr.length()+1) * sizeof(char) );
```
Comment (jmbr) [inline src/contrib/tweedledum/tweedledum.c]: Lines 27-29 can be replaced by `return strdup(resultstr.c_str());`

File: src/contrib/tweedledum/tweedledum.c [hunk 3]
```diff
@@ -14,6 +15,20 @@
 
 extern "C" {
 
+  extern char* tweedledum_synthesis_diagonal(double* angles, uint32_t size) {
+    using namespace tweedledum;
+    std::vector<double> vangles(angles, angles + size);
+    std::cout << vangles.size() << std::endl;
```
Comment (jmbr) [inline src/contrib/tweedledum/tweedledum.c]: Do we really want to output this? It seems to be for debugging purposes only.

Comment (notmgsk) [reply src/contrib/tweedledum/tweedledum.c]: Gone :)

File: src/contrib/tweedledum/tweedledum.c (hunk 2 above)
Comment (notmgsk) [reply src/contrib/tweedledum/tweedledum.c]: I remember thinking there must be a simpler way. Ty :)

**Comment by ecpeterson** [comment]
Recording / broadcasting some internal discussion:

The ideas around this PR—but not the PR itself—have two warts.

1. The implementation of this decomposition routine inside of `tweedledum` counts the position inside of the diagonal entries with an `int32`, and it detects the most significant nonzero bit in that position with a single-precision floating point logarithm. Together, these limit the number of qubits to which the routine applies. The routine itself doesn't actually look that complicated, and so there's some discussion of complaining to tweedledum's maintainers, filing a fix in their repository, rewriting it in Lisp to make use of its more flexible integer type, ... .

2. The gates to which this routine applies belong to the middle class in (uniformly controlled RZs == multiply-FORKED RZs) <= diagonal gates <= dense unitary operators. This routine also gives much better output than that of the dense decomposition `qs-compile.lisp`, but worse output than that of the FORKED compiler `ucr-compile.lisp`. Accordingly, the author of the diagonal compilation routine has to know to inject this compilation routine at a particular point in `*global-compilers*` — but a lot of the work in #210 was about automating these kinds of decisions in the case of a fixed qubit count, and it would be great to be able to extend that work to make this injection decision as well.

**Pull request closed without merging**